File: Robotran_py/userfctR/user_DrivenJoints.py
# -*- coding: utf-8 -*-
"""Module for the definition of driven joints."""
# Author: Robotran Team
# (c) Universite catholique de Louvain, 2020

def user_DrivenJoints(mbs_data, tsim):
    """Set the values of the driven joints directly in the MbsData structure.

    The position, velocity and acceleration of the driven joints must be set in
    the attributes mbs_data.q, mbs_data.qd and mbs_data.qdd .

    Parameters
    ----------
    mbs_data : MBsysPy.MbsData
        The multibody system associated to this computation.
    tsim : float
        The current time of the simulation.

    Returns
    -------
    None
    """
    # The motors are rigidly linked to the hip/thighs
    # The rotation of the hip is kept locked for the moment

    fixed_translations_ids = [mbs_data.joint_id["T_lh_fixed"], mbs_data.joint_id["T_rh_fixed"],
                              mbs_data.joint_id["T_lk_fixed"],mbs_data.joint_id["T_lk_fixed"],mbs_data.joint_id["R_hip"] ]
    for elem in fixed_translations_ids:
        mbs_data.q[elem]    = 0
        mbs_data.qd[elem]   = 0
        mbs_data.qdd[elem]  = 0  

    return

File: Robotran_c/userfctR/user_DrivenJoints.c
/* ---------------------------
 * Robotran - MBsysC
 * 
 * Template file for the driven joint function
 * 
 * (c) Universite catholique de Louvain
 *     
 */

 #include "math.h"
 #include "user_all_id.h"
 #include "mbs_data.h"
 
 
 void user_DrivenJoints(MbsData *mbs_data,double tsim)
 {
 
     int fixed_translations_ids[] = {
         T_lh_fixed_id,
         T_rh_fixed_id,
         T_lk_fixed_id,
         T_rk_fixed_id, 
         R_hip_id
     };
 
     int nb_ids = 5;
     for (int i = 0; i < nb_ids; i++)
     {
         int elem = fixed_translations_ids[i];
         mbs_data->q[elem] = 0.0;
         mbs_data->qd[elem] = 0.0;
         mbs_data->qdd[elem] = 0.0;
     }
 }